Detectors/CPV/calib/CPVCalibWorkflow/src/CPVPedestalCalibDevice.cxx

```diff
@@ -0,0 +1,253 @@
+    return;
+  }
+  // //TODO:
+  // //Get current map
+  // int nChanged=0;
+  // for(short i=o2::cpv::Geometry::kNCHANNELS; --i;){
+  //   short dp=mPedestals.getPedestal(i)-oldPed.getPedestal(i);
+  //   mPedDiff[i]=dp ;
+  //   if(abs(dp)>1){ //not a fluctuation
+  //     nChanged++;
+  //   }
+  // }
+  // if(nChanged>kMinorChange){ //serious change, do not update CCDB automatically, use "force" option to overwrite
+  //   mUpdateCCDB=false;
+  // }
+  // else{
+  //   mUpdateCCDB=true;
+  // }
+}
+
+//
+// Method to compare new pedestals and latest in ccdb
+// Do not update existing object automatically if difference is too strong
+// create object with validity range if far future (?) and send warning (e-mail?) to operator
+// if (mUseCCDB) { // read calibration objects from ccdb
+//    LHCphase lhcPhaseObjTmp;
+/*
+      // for now this part is not implemented; below, the sketch of how it should be done
+      if (mAttachToLHCphase) {
+        // if I want to take the LHCclockphase just produced, I need to loop over what the previous spec produces:
+        int nSlots = pc.inputs().getNofParts(0);
+        assert(pc.inputs().getNofParts(1) == nSlots);
+
+        int lhcphaseIndex = -1;
+        for (int isl = 0; isl < nSlots; isl++) {
+          const auto wrp = pc.inputs().get<CcdbObjectInfo*>("clbInfo", isl);
+          if (wrp->getStartValidityTimestamp() > tfcounter) { // replace tfcounter with the timestamp of the TF
+            lhxphaseIndex = isl - 1;
+            break;
+          }
+        }
+        if (lhcphaseIndex == -1) {
+          // no new object found, use CCDB
+	  auto lhcPhase = pc.inputs().get<LHCphase*>("tofccdbLHCphase");
+          lhcPhaseObjTmp = std::move(*lhcPhase);
+        }
+        else {
+          const auto pld = pc.inputs().get<gsl::span<char>>("clbPayload", lhcphaseIndex); // this is actually an image of TMemFile
+          // now i need to make a LHCphase object; Ruben suggested how, I did not try yet
+	  // ...
+        }
+      }
+*/
+
+o2::framework::DataProcessorSpec o2::cpv::getPedestalCalibSpec(bool useCCDB, bool forceUpdate, std::string path)
+{
+
+  std::vector<o2::framework::OutputSpec> outputs;
+  outputs.emplace_back("CPV", "PEDCALIBS", 0, o2::framework::Lifetime::Timeframe);
+  outputs.emplace_back("CPV", "PEDDIFF", 0, o2::framework::Lifetime::Timeframe);
+
+  return o2::framework::DataProcessorSpec{"PedestalCalibSpec",
+                                          o2::framework::select("A:CPV/RAWDATA"),
+                                          outputs,
+                                          o2::framework::adaptFromTask<CPVPedestalCalibDevice>(useCCDB, forceUpdate, path),
+                                          o2::framework::Options{}};
+}
+
+// //============================================================
+// DataProcessorSpec getTOFChannelCalibDeviceSpec(bool useCCDB){
+//   using device = o2::calibration::TOFChannelCalibDevice;
+//   using clbUtils = o2::calibration::Utils;
+
```

Comment:
this is just commented TOF code, I guess can be eliminated

Reply:
removed

Reply:
Dear @shahor02, @chiarazampolli, we will have two kinds of calibrations: pedestals (CPV) and pedestals, HG/LG ratio (PHOS) should process data from dedicated runs. Full run, usually relatively short ~few minutes should be processed. Another kind of calibrations: gains (CPV), time and gain (PHOS) will require several runs and even fills. In some discussions before we agreed that transient data should be stored on disk of DPL computers and processed when sufficient statistics will be collected. I tried to use corresponding lifetimes of devices. Do I miss something?

Reply:
We were speaking about the calibration info which is on the EPN accumulated (merged) in the containers, processed offline then deleted. These EPN containers, since there will be 250 of them,  will need to be aggregated, then the calibration can created. 
In case of your calib. devices, I am not sure: since you are using raw data, I guess they will be running on the FLP, since as far as understand, you will run the cell fits on FLPs and will not ship the raw data to EPN. I am not sure there we can accumulate data there (was never discussed), so, we may need to produce on FLP some intermediate data from raw, and send to EPN for accumulation.

Reply:
Calibrations like Pedestals and HG/LG ratio do use raw data. Their output histos (in addition to CCDB objects) will be used for bad map and should be shipped to EPN. There are light-weight devices and should not consume a lot of CPU. Even if they will interfere with data taking, I think this is not a problem if part of pedestal or LED runs will be lost. Another option: split device to standard raw decoding part on FLP and analysis part on EPN. 
PHOS time and gain device will use clusters and reside on EPN. CPV gain device can use raw data as now or digits if we decide to move it to EPN. We will discuss this with CPV experts.